Adding Swift MapKit Page (#368)
learning-software-engineering/learning-software-engineering.github.io: Topics/Tech_Stacks/Swift_Mapkit.md

# Introduction to MapKit for SwiftUI

> SwiftUI is a very vast collection of features designed by Apple to make iOS development…swift-er. One of those handy features is it’s ability to make maps, specifically with MapKit. 

Reviewer: teehee!

**Requirements:** To use these MapKit features, you need to use Xcode 15 and later.

## Table of contents
### [What is MapKit?](#what-is-mapkit-1)
### [Creating a map](#creating-a-map-1)
### [Markers](#markers-1)
### [Annotations](#annotations-1)
### [Camera Position](#camera-position-1)
### [Additional Resources](#additional-resources-1)

## What is MapKit?

> MapKit is one of Apple’s API framework that allows iOS developers to build map-centered views and apps easily and efficiently. 

Reviewer: framework(s) needs to be plural

Based on the [Apple documentation](https://developer.apple.com/documentation/mapkit/), MapKit can:

- Embed maps directly into your app’s windows and views.
- Add annotations and overlays to a map to call out points of interest.
- Add LookAround capabilities to enable users to explore locations at street level.
- Respond to user interactions with well known points of interest, geographical features, and boundaries.

> - Provide text completion to make it easy for users to search for a destination or point of interest.

Reviewer: text completion in what context? i'm assuming in a search bar if you're building a navigation app?

Author: This is straight from the apple documentation! Maybe I can add some examples as to when you would use these

Reviewer: That makes sense! I don't think it was super clear that this was pulled from their docs, so maybe add a short line?

Author: But yes you are correct.

Author: I have a line before hand that says based off the apple documentation with link then list the features. Do you think that I should add more?

Reviewer: Apparently I can't read! You are absolutely correct and it's quite clear in its current state. This looks good to go!

## Creating a Map
Let’s start with a basic map.
<p align="center">
  <img width=1000 src="https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/113125436/15f46149-bc31-4b53-98d5-0e67b44abb26"/>
</p>

To create a basic map, you first need to import MapKit into your Swift file. Then simply call Map() in the view. 

```swift
import SwiftUI
import MapKit

struct MapContentView: View {
    var body: some View {       
            Map()
    }
}

#Preview {
    MapContentView()
}
```

> Congratulations you made a map! As you can tell this is just the bare bones. Now that you have a map, you have to customize it to fit your needs.

Reviewer: ```Congratulations, you made a map!
```

> Congratulations you made a map! As you can tell this is just the bare bones. Now that you have a map, you have to customize it to fit your needs.

Reviewer: ```suggestion
Congratulations, you made a map!
```

Author: You are so right! Thank you for catching the comma.

### Customizing your map

> **Map Style**

Reviewer: ```suggestion
**Map Style**
```
Needs another inline after the title, it isn't displaying properly in the HTML render

  ```

**Icon**

  To change the what is displayed in the Marker bubble you can add the `systemImage`, `image`, or `monogram` parameters to your Marker as shown below.

  System Image:
  ```swift
     Marker("Bahen", systemImage: "star.fill", coordinate: bahen)
  ```
   Image:
   ```swift
     Marker("Robarts", image: "<Your Image Name>", coordinate: robarts)
   ```

  Monogram
  ```swift
     Marker("University College", monogram: Text("UC"), coordinate: uc)
  ```

If you want more freedom with your designing map pins, try using [Annotations](#annotations) instead.
## Annotations
Annotations allow for MapKit developrs to completely redesign the location coordinate indicators.
Unlike markers which had a set amount of parameters and there was no way to escape the bubble, Annotations instead takes in a Swift UI `View()`.

For example, here we made the annotation to just be a system image with padding rather than a marker bubble.
```swift
import SwiftUI
import MapKit

struct MapContentView: View {
    
    let robarts = CLLocationCoordinate2D(latitude: 43.664486, longitude: -79.399689)
    
    var body: some View {
        Map(){

>             Annotation("Robarts", coordinate: robarts){

Reviewer: Indentations here look a little strange in the richtext/rendered format, but not sure if it's fixable? They look correct in the markdown file...

Author: Hmmmm, interesting... I will look into this and try to fix it.

                Image(systemName: "books.vertical")
                    .padding(6)
                    .foregroundColor(.white)
                    .background(/*@START_MENU_TOKEN@*/.blue/*@END_MENU_TOKEN@*/)
            }
            
        }
    }
}

#Preview {
    MapContentView()
}
```
<p align="center">
<img width=1000 src="https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/113125436/c64ff5e5-8286-4ad4-b24f-3eae8545f247"/>
</p>

Overall, the annotations allow you to make the map pins compeletely your own.

## Camera Position

> By default the map focuses on the map contents (markers, annotaions, etc). To focus on a particular region or location, we can use `MapCameraPosition`. This can be helpful in a couple cases such as:

Reviewer: ```suggestion
By default the map focuses on the map contents (markers, annotations, etc). To focus on a particular region or location, we can use `MapCameraPosition`. This can be helpful in a couple cases such as:
```

Author: Wow I have so many typos

```swift
@State private var cameraPosition: MapCameraPosition = .automatic
```

An example of using `MapCameraPosition` to focus on a location is setting `.region`. In the example below, we focus the map on the CN Tower even though the marker is at Robarts.
```swift
import SwiftUI
import MapKit

struct MapContentView: View {
    let robarts = CLLocationCoordinate2D(latitude: 43.664486, longitude: -79.399689)
    @State private var cameraPosition: MapCameraPosition = .region(MKCoordinateRegion(
        center: .init(latitude: 43.6426, longitude: -79.3871),
        span: .init(latitudeDelta: 0.01, longitudeDelta: 0.01))
    )

    
    var body: some View {
        Map(position: $cameraPosition){
            Marker("Robarts", coordinate: robarts)
            
        }
    }
}

#Preview {
    MapContentView()
}
```
<p align="center">
<img width=1000 src="https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/113125436/931abbb0-7217-4e86-84e8-3bec30437d47"/>
</p>

Beyond these introductory examples, `MapCameraPosition` has a big potential and is very useful when creating dynamic maps.

## Additional Resources

> MapKit for SwiftUI is very large and has a lot of possible implementations, so this tutorial was just the tip of the ice berg! Below are some additional resources

Reviewer: ```suggestion
MapKit for SwiftUI is very large and has a lot of possible implementations, so this tutorial was just the tip of the iceberg! Below are some additional resources
```
